### PR TITLE
feat(auth): Add support for Email MFA

### PR DESCRIPTION
**Pull request by vincetran**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:* N/A

*Description of changes:*
The PR aims to add support for Email MFA that has been released by Cognito.
https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pool-settings-advanced-security-email-mfa.html

Draft PR for the Authenticator UI change: https://github.com/aws-amplify/amplify-ui-android/pull/199

*How did you test these changes?*
Manual testing, new and existing unit tests pass, new and existing integration tests pass

*Documentation update required?*
- [ ] No
- [x] Yes (Please include a PR link for the documentation update) - PR Link Pending

*General Checklist*
- [x] Added Unit Tests
- [x] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    companion object {
+        private var authPlugin = AWSCognitoAuthPlugin()
+        private var apiPlugin = AWSApiPlugin()
+        lateinit var synchronousAuth: SynchronousAuth
+        var mfaCode = ""
+        var abortableLatch: AbortableCountdownLatch? = null
+
+        @JvmStatic
+        @BeforeClass
+        fun initializePlugin() {
```
Comment (tylerjroach) [inline aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt:60]: Can we instead start cleanly each time? We've made efforts to be able to do so. This should work as long as no code is calling an `Amplify` singleton.

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+import kotlin.test.assertFailsWith
+import org.junit.After
+import org.junit.BeforeClass
+import org.junit.Test
+
+class AWSCognitoAuthPluginEmailMFATests {
+
+    private val password = "${UUID.randomUUID()}BleepBloop1234!"
+    private val userName = "test${Random().nextInt()}"
+    private val email = "$userName@testdomain.com"
```
Comment (tylerjroach) [inline aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt:49]: are we sending real verification emails to testdomain.com? This needs to be changed.

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/testutils/AbortableCountdownLatch.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.auth.cognito.testutils
+
+import java.util.concurrent.CountDownLatch
+
+class AbortableCountdownLatch(count: Int) : CountDownLatch(count) {
```
Comment (tylerjroach) [inline aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/testutils/AbortableCountdownLatch.kt:19]: I don't understand the purpose of this class. Why are we not using a normal CountDownLatch?

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+import java.util.Random
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+import org.junit.After
+import org.junit.BeforeClass
+import org.junit.Test
+
+class AWSCognitoAuthPluginEmailMFATests {
```
Comment (tylerjroach) [inline aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt:45]: Can you explain here how you are getting an email mfa code by sending to an invalid email?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/helpers/MFAHelper.kt [hunk 5]
```diff
@@ -16,22 +16,69 @@
...
+        getAllowedMFASetupTypesFromChallengeParameters(challenge.parameters) == setOf(MFAType.EMAIL)
+
+internal fun getAllowedMFATypesFromChallengeParameters(challengeParameters: Map<String, String>?): Set<MFAType> {
+    val mfasCanChoose = challengeParameters?.get("MFAS_CAN_CHOOSE") ?: return emptySet()
+    val result = mutableSetOf<MFAType>()
+    mfasCanChoose.replace(Regex("\\[|\\]|\""), "").split(",").forEach {
+        when (it) {
+            "SMS_MFA" -> result.add(MFAType.SMS)
+            "SOFTWARE_TOKEN_MFA" -> result.add(MFAType.TOTP)
+            "EMAIL_MFA" -> result.add(MFAType.EMAIL)
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/helpers/MFAHelper.kt:65]: Confirmed this should be EMAIL_OTP. We are returning the wrong value to the customer, which is why nothing currently "appears" broken.

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt (hunk 4 above)
Comment (tylerjroach) [reply aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt:45]: Harshdeep explained this to me. Please just look at changing the email shown here.

File: aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt (hunk 2 above)
Comment (vincetran) [reply aws-auth-cognito/src/androidTest/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPluginEmailMFATests.kt]: Using the same email as Swift

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**